components/logins/src/migrate_sqlcipher_db.rs

```diff
@@ -67,40 +99,95 @@ fn sqlcipher_3_compat(conn: &Connection) -> Result<()> {
     Ok(())
 }
 
-fn sqlcipher_export(conn: &mut Connection, new_db_path: impl AsRef<Path>) -> Result<()> {
-    // Export sqlite data to a plaintext DB using the strategy from example 2 here:
-    // https://www.zetetic.net/sqlcipher/sqlcipher-api/index.html#sqlcipher_export
-    let path = new_db_path.as_ref().as_os_str();
-    let path = path
-        .to_str()
-        .ok_or_else(|| ErrorKind::InvalidPath(path.to_os_string()))?;
-
-    conn.execute("ATTACH DATABASE ? AS plaintext key ''", &[path])?;
-    // this one is a bit weird because it's a SELECT statement that we know will return 0 rows.
-    // rusqlite will return an Error if we use execute(), so we use query_row with a dummy closure
-    conn.query_row("SELECT sqlcipher_export('plaintext')", NO_PARAMS, |_| Ok(0))?;
-    conn.execute("DETACH DATABASE plaintext", NO_PARAMS)?;
-    Ok(())
+//Manually copy over row by row from sqlcipher db to a plaintext db
+pub fn migrate_from_sqlcipher_db(
+    cipher_conn: &mut Connection,
+    new_db_store: LoginStore,
+    encryption_key: &str,
+) -> Result<MigrationMetrics> {
+    let start_time = Instant::now();
+
+    let mut metrics: MigrationMetrics = MigrationMetrics::default();
+
+    // Select From both LoginsL and LoginsM with a union to ensure we're covering our
+    // migration cases (one table has but not the other, etc)
+    let mut select_stmt = cipher_conn.prepare(&GET_BY_GUID_SQL)?;
+
+    // Use raw rows to avoid extra copying since we're looping over an entire table
+    let mut rows = select_stmt.query(NO_PARAMS)?;
+    while let Some(row) = rows.next()? {
+        metrics.num_processed += 1;
+        let guid: String = row.get("guid")?;
+        let username: String = row.get("username").unwrap_or_default();
+        let password: String = row.get("password")?;
+        // migrating hostname to the new column origin
+        let origin: String = row.get("hostname")?;
+        let http_realm: Option<String> = row.get("httpRealm")?;
+        // migrating formSubmitURL to the new column action origin
+        let form_action_origin: Option<String> = row.get("formSubmitURL")?;
+        let username_field: Option<String> = row.get("usernameField")?;
+        let password_field: Option<String> = row.get("passwordField")?;
+        let time_created: i64 = row.get("timeCreated")?;
+        let time_last_used: i64 = row.get("timeLastUsed").unwrap_or_default();
+        let time_password_changed: i64 = row.get("timePasswordChanged")?;
+        let times_used: i64 = row.get("timesUsed")?;
+
+        // TODO: Discuss
+        // Need to handle in loginsL: local_modified, is_deleted
+        // Feels like we potentially shouldn't migrate these fields?
+
+        // Need to handle in loginsM: is_overridden, server_modified
+        // Similar to the other: I only see server_modified potentially being needed
```

Comment:
I think we need to copy all of the columns since they're all used with the current code.  We discussed updating the schema to make it match the other components.  If we did that, these columns might not be needed anymore but it would require changing the syncing code.  And I'd like to avoid adding another large change in the branch if we can help it.